File: Easy2D_MATLAB/PREP.m
function [fid2,NNODE, NELEM, X, Y, NODE, KIND, TEMP, CA, CB, CC,FREC,Field, Exterior,Px,Py,VINF]= PREP(inputfilename)
%
%---- PREPROCESSOR
%
% clc
% clear
% inputfilename='TestCase2_2.dat';
fid = fopen(inputfilename,'r');
fid2= fopen('output.dat','w');
InputText=textscan(fid,'%s',150,'delimiter','\n'); % Read strings delimited by a carriage return
BLOCK=InputText{1};
NNN=str2num(BLOCK{3});
NNODE=NNN(1);
NREC=NNN(2);
EEE=str2num(BLOCK{5+NREC});
NELEM=EEE(1);
EREC=EEE(2);
BREC=str2num(BLOCK{7+NREC+EREC});
Field=0;
Exterior=0;
FREC=0;
if strcmp(BLOCK{2},'NODES') ==0
    disp('Error on NODES Command');
    return
end
if strcmp(BLOCK{4+NREC},'ELEMENTS')==0
    disp('Error on ELEMENTS Command');
    return
end
if strcmp(BLOCK{6+NREC+EREC},'BOUNDARY CONDITIONS')==0
    disp('Error on BOUNDARY CONDITIONS Command');
    return
end
if strcmp(BLOCK{8+NREC+EREC+BREC},'SOLVE')==0
    if strcmp(BLOCK{8+NREC+EREC+BREC},'FIELD')==1
        FREC=str2num(BLOCK{9+NREC+EREC+BREC});
        Field=2;
        Exterior=0;
        if (strcmp(BLOCK{10+NREC+EREC+BREC+FREC},'EXTERIOR')==1)
            if(strcmp(BLOCK{11+NREC+EREC+BREC+FREC},'VINF')==1)
                VINF= str2num(BLOCK{12+NREC+EREC+BREC+FREC});
            end
            Exterior=3;
        end
    else
        if (strcmp(BLOCK{8+NREC+EREC+BREC},'EXTERIOR')==1)
            if (strcmp(BLOCK{9+NREC+EREC+BREC},'VINF')==1)
                VINF=str2num(BLOCK{10+NREC+EREC+BREC});
            end
            Field=0;
            Exterior=3;
        end
    end
    
end
if Exterior==0;
    VINF=0;
end
if strcmp(BLOCK{8+NREC+EREC+BREC+FREC+Field+Exterior},'SOLVE')==0
    disp('Couldnt Catch SOLVE Command, Solution Terminated');
    return
end

fprintf(fid2,'%s \n \n', BLOCK{1});
nodes=zeros(NNN(2),6);
for i=4:NREC+3
    nodes(i-3,:)= str2num(BLOCK{i});
end
N1=nodes(:,1);
N2=nodes(:,2);
X1=nodes(:,3);
Y1=nodes(:,4);
X2=nodes(:,5);
Y2=nodes(:,6);
[X,Y]=Rnode(fid2,NNODE,NREC,N1,N2,X1,Y1,X2,Y2);
elements=zeros(EREC,4);
for j=6+NREC:6+NREC+EREC-1
    elements(j-5-NREC,:)=str2num(BLOCK{j});
end
NODEI=elements(:,1);
NODEL=elements(:,2);
NUMBER=elements(:,3);
KINDI=elements(:,4);
[NODE, KIND]=Relem(fid2,NNODE,NELEM,EREC,NODEI,NODEL,NUMBER,KINDI);
bcs=zeros(BREC,6);
for k=8+NREC+EREC:8+NREC+EREC+BREC-1
    bcs(k-7-NREC-EREC,:)=str2num(BLOCK{k});
end
K1=bcs(:,1);
K2=bcs(:,2);
NOD=bcs(:,3);
CA1=bcs(:,4);
CB1=bcs(:,5);
CC1=bcs(:,6);
[TEMP,CA,CB,CC]= BC(fid2, NNODE, NELEM, NODE, KIND, BREC,K1,K2,NOD,CA1,CB1,CC1);
if Field==2
    FIELDS=zeros(FREC,2);
    for l=10+NREC+EREC+BREC:10+NREC+EREC+BREC+FREC-1
        FIELDS(l-9-NREC-EREC-BREC,:)=str2num(BLOCK{l});
    end
    Px=FIELDS(:,1);
    Py=FIELDS(:,2);
end
fclose(fid);
end

File: PREP.py
import numpy as np
from Rnode import Rnode
from Relem import Relem
from BC import BC


def PREP(inputfilename):
    """
    PREPROCESSOR
    """

    # inputfilename='TestCase2_2.dat';
    inputText = ""
    with open(inputfilename, 'r') as inputFile:
        inputText = inputFile.read().splitlines()
    outputFile = open('output.dat', 'w')

    # Check for required commands and grab index
    try:
        NODES_idx = inputText.index('NODES')
        # Grabs number of nodes and records from input file
        NNN = np.fromstring(inputText[NODES_idx+1], dtype=int, sep=',')
        NNODE = NNN[0]  # Gets number of nodes
        NREC = NNN[1]  # Gets number of records
    except Exception:
        print('Error on NODES Command')
        return

    try:
        ELEM_idx = inputText.index('ELEMENTS')
        # Gets number of elements and records
        EEE = np.fromstring(inputText[ELEM_idx+1], dtype=int, sep=',')
        NELEM = EEE[0]  # Number of elements
        EREC = EEE[1]   # Number of element records
    except Exception:
        print('Error on ELEMENTS Command')
        return

    try:
        BC_idx = inputText.index('BOUNDARY CONDITIONS')
        # Gets number of records for BCs
        BREC = np.fromstring(inputText[BC_idx+1], dtype=int, sep=',')[0]
    except Exception:
        print('Error on BOUNDARY CONDITIONS Command')
        return

    try:
        SOLVE_idx = inputText.index('SOLVE')
    except Exception:
        print('Error on SOLVE Command')
        return

    Field = 0
    Exterior = 0
    FREC = 0
    try:
        FIELD_idx = inputText.index('FIELD')
        FREC = np.fromstring(inputText[FIELD_idx+1], dtype=int, sep=',')[0]
        Field = 2
    except Exception:
        pass

    if 'EXTERIOR' in inputText:
        Exterior = 3
        if 'VINF' in inputText:
            VINF = np.fromstring(inputText[inputText.index('VINF')+1],
                                 dtype=int, sep=',')[0]

    outputFile.write('{}\n\n'.format(inputText[0]))
    nodes = np.zeros((NREC, 6))
    for i, node in enumerate(inputText[NODES_idx+2:NODES_idx+NREC+2]):
        nodes[i, :] = np.fromstring(node, sep=',')

    N1 = nodes[:, 0]
    N2 = nodes[:, 1]
    X1 = nodes[:, 2]
    Y1 = nodes[:, 3]
    X2 = nodes[:, 4]
    Y2 = nodes[:, 5]
    [X, Y] = Rnode(outputFile, NNODE, NREC, N1, N2, X1, Y1, X2, Y2)

    elements = np.zeros((EREC, 4))
    for j, elem in enumerate(inputText[ELEM_idx+2:ELEM_idx+EREC+2]):
        elements[j, :] = np.fromstring(elem, sep=',')

    NODEI = elements[:, 0].astype(int)
    NODEL = elements[:, 1].astype(int)
    NUMBER = elements[:, 2].astype(int)
    KINDI = elements[:, 3].astype(int)
    [NODE, KIND] = Relem(outputFile, NNODE, NELEM, EREC, NODEI, NODEL, NUMBER,
                         KINDI)

    bcs = np.zeros((BREC, 6))
    for k, bc in enumerate(inputText[BC_idx + 2:BC_idx+BREC+2]):
        bcs[k, :] = np.fromstring(bc, sep=',')

    K1 = bcs[:, 0].astype(int) - 1
    K2 = bcs[:, 1].astype(int) - 1
    NOD = bcs[:, 2].astype(int)
    CA1 = bcs[:, 3]
    CB1 = bcs[:, 4]
    CC1 = bcs[:, 5]
    [TEMP, CA, CB, CC] = BC(outputFile, NNODE, NELEM, NODE, KIND, BREC, K1, K2,
                            NOD, CA1, CB1, CC1)

    if Field == 2:
        FIELDS = np.zeros((FREC, 2))
        for l, FIELD in enumerate(inputText[FIELD_idx+2:FIELD_idx+FREC+2]):
            FIELDS[l, :] = np.fromstring(FIELD, sep=',')

        Px = FIELDS[:, 0]
        Py = FIELDS[:, 1]

    return [outputFile, NNODE, NELEM, X, Y, NODE, KIND, TEMP, CA, CB, CC, FREC,
            Field, Exterior, Px, Py, VINF]
